storage/executable-store/src/lib.rs

```diff
@@ -0,0 +1,365 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use aptos_crypto::HashValue;
+use aptos_infallible::Mutex;
+use aptos_logger::error;
+use aptos_types::executable::Executable;
+use dashmap::DashMap;
+use num_derive::FromPrimitive;
+use std::{
+    hash::Hash,
+    sync::atomic::{AtomicU8, AtomicUsize, Ordering},
+};
+
+/// Represents the state of the ExecutableStore. The goal is to catch errors in intended
+/// use, and reset the executable cache to an empty state if its state can't guarantee
+/// matching the state after the parent block id.
+///
+/// During the execution, the cache might get updated, i.e. if a new executable at storage
+/// version gets stored. After block execution, more updates may occur to align the cache
+/// contents to the new boundary: published modules may invalidated previously stored
+/// executables, and new corresponding executables may also be available. 'mark_updated'
+/// method can be called afterwards to mark state accordingly. Missing Updated state
+/// would indicate that the cache wasn't sychronized to the new block, and should not be
+/// re-used for the subsequent blocks. The Void state in such cases allows graceful
+/// handling by resetting / clearing the cache as needed.
+///
+/// Finally, ExecutableStore must be pruned, to control its size, and then the hash of the
+/// block that was executed needs to be set, marking it full circle in the Ready state.
+#[derive(Copy, Clone, FromPrimitive, PartialEq, Debug)]
+enum ExecutableStoreState {
+    Ready = 0,   // The cache is ready-to-use at a state after block id is recorded.
+    Updated = 1, // The cache is updated / synchronized after the block execution.
+    Pruned = 2,  // The cache is pruned (must record executed block id to become ready).
+    Void = 3,    // Unexpected state update order (expected: Ready -> Updated -> Pruned).
+}
+
+pub struct ExecutableStore<K: Eq + Hash, X: Executable> {
+    executables: DashMap<K, X>,
+    total_size: AtomicUsize,
+    state: AtomicU8,
+    block_id: Mutex<Option<HashValue>>,
```

Comment:
@sasha8 discussion: Make HashValue generic so in the future we could work with other executors (e.g. chunk id).